File: src/main/java/io/github/tanguygab/tabadditions/shared/features/ConditionalAppearance.java
package io.github.tanguygab.tabadditions.shared.features;

import io.github.tanguygab.tabadditions.shared.features.advancedconditions.AdvancedConditions;
import lombok.Getter;
import me.neznamy.tab.shared.Property;
import me.neznamy.tab.shared.TAB;
import me.neznamy.tab.shared.features.types.*;
import me.neznamy.tab.shared.platform.TabPlayer;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;

public class ConditionalAppearance extends RefreshableFeature implements JoinListener, QuitListener, UnLoadable {

    @Getter private final String featureName = "Conditional Appearance";
    @Getter private final String refreshDisplayName = "&aConditional Appearance&r";
    private final TAB tab;
    private final Plugin plugin;
    private final boolean def;
    private final boolean pwp;
    private final Map<TabPlayer, Property> properties = new HashMap<>();

    public ConditionalAppearance(Object plugin, boolean def) {
        tab = TAB.getInstance();
        this.plugin = (Plugin) plugin;
        this.def = def;
        pwp = tab.getFeatureManager().isFeatureEnabled("PerWorldPlayerList");
        if (pwp) addUsedPlaceholder("%world%");
        for (TabPlayer p : tab.getOnlinePlayers()) onJoin(p);
    }

    @Override
    public void onJoin(@NotNull TabPlayer player) {
        properties.put(player, player.loadPropertyFromConfig(this,"appearance-condition", ""));
        refresh(player,true);
    }

    @Override
    public void onQuit(@NotNull TabPlayer player) {
        properties.remove(player);
    }

    @Override
    public void refresh(@NotNull TabPlayer p, boolean force) {
        for (TabPlayer all : tab.getOnlinePlayers()) {
            if (p == all) return;
            refresh(p, all);
            refresh(all, p);
        }
    }

    private void refresh(TabPlayer target, TabPlayer viewer) {
        if (getCondition(target,viewer)) sync(()->show(p(viewer),p(target)));
        else sync(()->hide(p(viewer),p(target)));
    }

    private void sync(Runnable run) {
        plugin.getServer().getScheduler().runTask(plugin,run);
    }

    private Player p(TabPlayer p) {
        return (Player) p.getPlayer();
    }

    public boolean getCondition(TabPlayer target, TabPlayer viewer) {
        if (target == null || viewer == null) return def;
        if (pwp && !target.world.equals(viewer.world)) return def;
        Property prop = properties.get(target);
        if (prop == null) return def;
        String cond = prop.getCurrentRawValue();
        if (cond.isEmpty()) return def;
        return def != AdvancedConditions.getCondition(cond).isMet(viewer,target);
    }

    @Override
    public void unload() {
        for (Player p : Bukkit.getOnlinePlayers()) {
            for (Player all : Bukkit.getOnlinePlayers()) {
                if (p != all) show(p,all);
            }
        }
    }

    @SuppressWarnings("deprecation")
    private void show(Player viewer, Player target) {
        try {viewer.showPlayer(plugin, target);}
        catch (NoSuchMethodError e) {viewer.showPlayer(target);}
    }
    @SuppressWarnings("deprecation")
    private void hide(Player viewer, Player target) {
        try {viewer.hidePlayer(plugin, target);}
        catch (NoSuchMethodError e) {viewer.hidePlayer(target);}
    }

}

File: src/main/kotlin/io/github/tanguygab/tabadditions/shared/features/ConditionalAppearance.kt
package io.github.tanguygab.tabadditions.shared.features

import io.github.tanguygab.tabadditions.shared.features.advancedconditions.AdvancedConditions
import me.neznamy.tab.shared.Property
import me.neznamy.tab.shared.TAB
import me.neznamy.tab.shared.features.types.JoinListener
import me.neznamy.tab.shared.features.types.QuitListener
import me.neznamy.tab.shared.features.types.RefreshableFeature
import me.neznamy.tab.shared.features.types.UnLoadable
import me.neznamy.tab.shared.platform.TabPlayer
import org.bukkit.entity.Player
import org.bukkit.plugin.Plugin

class ConditionalAppearance(plugin: Any?, private val def: Boolean)
    : RefreshableFeature(), JoinListener, QuitListener, UnLoadable {
    private val tab = TAB.getInstance()
    private val plugin = plugin as Plugin
    private val pwp = tab.featureManager.isFeatureEnabled("PerWorldPlayerList")
    private val properties = mutableMapOf<TabPlayer, Property>()

    init {
        if (pwp) addUsedPlaceholder("%world%")
        tab.onlinePlayers.forEach { onJoin(it) }
    }

    override fun getFeatureName() = "Conditional Appearance"
    override fun getRefreshDisplayName() = "&aConditional Appearance&r"

    override fun onJoin(player: TabPlayer) {
        properties[player] = player.loadPropertyFromConfig(this, "appearance-condition", "")
        refresh(player, true)
    }

    override fun onQuit(player: TabPlayer) {
        properties.remove(player)
    }

    override fun refresh(player: TabPlayer, force: Boolean) {
        tab.onlinePlayers.forEach {
            if (player === it) return
            refresh(player, it)
            refresh(it, player)
        }
    }

    private fun refresh(target: TabPlayer, viewer: TabPlayer) {
        val function = if (getCondition(target, viewer)) ::show else ::hide
        sync { function(p(viewer), p(target)) }
    }

    private fun sync(run: Runnable) {
        plugin.server.scheduler.runTask(plugin, run)
    }

    private fun p(p: TabPlayer) = p.player as Player

    fun getCondition(target: TabPlayer, viewer: TabPlayer): Boolean {
        if (pwp && target.world != viewer.world) return def

        val prop = properties[target] ?: return def
        val cond = prop.currentRawValue
        if (cond.isEmpty()) return def
        return def != AdvancedConditions.getCondition(cond)!!.isMet(viewer, target)
    }

    override fun unload() {
        plugin.server.onlinePlayers.forEach { viewer ->
            plugin.server.onlinePlayers.forEach { target ->
                if (viewer != target) show(viewer, target)
            }
        }
    }

    @Suppress("DEPRECATION")
    private fun show(viewer: Player, target: Player) {
        try {
            viewer.showPlayer(plugin, target)
        } catch (_: NoSuchMethodError) {
            viewer.showPlayer(target)
        }
    }

    @Suppress("DEPRECATION")
    private fun hide(viewer: Player, target: Player) {
        try {
            viewer.hidePlayer(plugin, target)
        } catch (_: NoSuchMethodError) {
            viewer.hidePlayer(target)
        }
    }
}
